### PR TITLE
Issue #1 | Basic linalg classes and routines, Normal rand

### LINKED ISSUE #1: Support basic features in linalg and rand

### PR DESCRIPTION
**Pull request by Mr-S-Mirzoev**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,16 @@
...
+
+# Install dir
+install
+cmake
+
+# MacOS specific
+*.DS_Store
+
+# VSCode specific
+*.vscode
```
Comment (Mr-S-Mirzoev) [inline .gitignore]: trailing newline

File: install_deps.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash -xe
+
+SCRIPT_DIR=$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)
+
+if [ ! -d "${SCRIPT_DIR}/install/eigen" ]; then
```
Comment (Mr-S-Mirzoev) [inline install_deps.sh]: function, instead of repeating code

File: source/modules/linalg/inc/linalg/vector.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    std::ostream& operator<<(std::ostream& os, const Vector& v);
+    Vector ones(size_t n);
+    Vector zeros(size_t n);
+    Vector range(double begin, double end, size_t size);
+    Vector interp(Vector const& xvals, Vector const& yvals, Vector const& x);
+    double interp(Vector const& xvals, Vector const& yvals, double x);
+} // namespace linalg
+
+// #include "../vector_impl.h"
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/inc/linalg/vector.h]: trailing newspace

File: source/modules/linalg/inc/matrix_impl.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// namespace linalg
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/inc/matrix_impl.h]: remove

File: source/modules/linalg/inc/vector_impl.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// namespace linalg
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/inc/vector_impl.h]: remove

File: source/modules/linalg/src/algorithms.cpp [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#include "linalg/algorithms.h"
+#include "linalg/matrix.h"
+
+#include <Eigen/Core>
+#include <Eigen/Cholesky>
+
+namespace linalg
+{
+Matrix algorithms::cholesky(Matrix const& A)
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/src/algorithms.cpp:9]: add comments

File: source/modules/linalg/src/matrix.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    return data_;
+}
+
+Matrix Matrix::T() const
+{
+    Eigen::MatrixXd data_T = data_.transpose();
+    return data_T;
+}
+
+// Vector& Matrix::operator()(size_t row_idx)
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/src/matrix.cpp]: remove

File: source/modules/plotter/src/plotter.cpp [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+{
+    using namespace matplot;
+
+    std::vector<double> x{v.as_std_vector()};
+
+    auto h = hist(x);
+    std::cout << "Histogram with " << h->num_bins() << " bins" << std::endl;
+
+    show();
+}
```
Comment (Mr-S-Mirzoev) [inline source/modules/plotter/src/plotter.cpp]: trailing newline

File: source/modules/rnd/src/random.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    std::default_random_engine generator_(get_random_seed());
+    std::poisson_distribution<int> distribution_(mean);
+
+    linalg::Vector out(size);
+
+    for (int counter_(0); counter_ < size; ++counter_)
+        out(counter_) = distribution_(generator_);
+
+    return out;
+}
```
Comment (Mr-S-Mirzoev) [inline source/modules/rnd/src/random.cpp]: trailing newline

**Review by Mr-S-Mirzoev — COMMENTED** [review]
To fix:

File: source/modules/linalg/inc/linalg/matrix.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        Eigen::MatrixXd data_;
+
+        Matrix(Eigen::MatrixXd const& data);
+        Eigen::MatrixXd const& get_data() const;
+        Eigen::MatrixXd& get_data_writable();
+
+    public:
+        Matrix(int m = 0, int n = 0);
+
+        Matrix T() const;
```
Comment (MaksimSerebryakov) [inline source/modules/linalg/inc/linalg/matrix.h:27]: what do you think about renaming this method?
it was a bit difficult for me to understand the meaning before checking matrix.cpp
maybe smth with "transpose" or if T is common for matrix, it's okey :)

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required (VERSION 3.0)
+project (num_stoch_calc)
+
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "Default C++ standard")
+set(NUM_STOCH_CALC_EXAMPLE_EXE num_stoch_calc_exe)
+set(NUM_STOCH_CALC_ROOT ${CMAKE_CURRENT_LIST_DIR})
+
+add_executable (${NUM_STOCH_CALC_EXAMPLE_EXE} source/main.cpp)
+
+set(MODULE_LINALG_NAME linalg)
```
Comment (Mr-S-Mirzoev) [inline CMakeLists.txt]: investigate if it's possible to wrap it in function

File: install_deps.sh [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        $4;
+
+        echo "$2 library installation complete"
+    else
+        echo "$2 library already installed, skipping installation"
+    fi
+}
+
+build_dep eigen Eigen eigen_generate_project eigen_post_install_command
+build_dep matplotpp "Matplot++" mpp_generate_project
```
Comment (Mr-S-Mirzoev) [inline install_deps.sh]: trailing newline

File: source/main.cpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        // Incremental simulation with linear interpolation
+        WInt[i] = brownian_motion::incremental_interpolation(t, T, N[i]);
+
+        // Exact simulation over Brownian bridge
+        W[i] = brownian_motion::bridge_interpolation(t, T, N[i]);
+    }
+
+    plotter::plot_details config;
+    config.LineWIdth = 1.5;
+    config.XLabel = "$t$";
```
Comment (Mr-S-Mirzoev) [inline source/main.cpp]: investigate why latex isn't rendered. Possibly $ are mismatched

File: source/modules/linalg/inc/linalg/matrix.h [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+        // Vector& operator()(size_t row_idx);
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/inc/linalg/matrix.h]: remove

File: source/modules/linalg/inc/linalg/matrix.h [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        double& operator()(size_t row_idx, size_t col_idx);
+        // Vector const& operator()(size_t row_idx) const;
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/inc/linalg/matrix.h]: remove

File: source/modules/linalg/src/spline.cpp [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    // x values need to be scaled down in extraction as well.
+    return spline_(scaled_value(x))(0);
+}
+
+Vector SplineFunction::operator()(Vector const& x) const
+{
+    return Vector{x.get_data().unaryExpr([this](double x) { return this->operator()(x); })};
+}
+
+// Helpers to scale X values down to [0, 1]
```
Comment (Mr-S-Mirzoev) [inline source/modules/linalg/src/spline.cpp]: move doc-string to header

File: source/modules/plotter/inc/plotter_impl.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+template <typename... Types>
+void plotter::show_plot(plotter::plot_details const& config, Types const&... args)
+{
+    using namespace matplot;
+
+    const std::size_t nargs = sizeof...(args) / 2;
+
+    auto p = plot((args.as_std_vector(), ...));
+
+    // if (config.Legend.has_value())
```
Comment (Mr-S-Mirzoev) [inline source/modules/plotter/inc/plotter_impl.h]: remove

File: source/modules/plotter/inc/plotter_impl.h [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        xlabel(config.XLabel.value());
+    }
+
+    if (config.YLabel.has_value())
+    {
+        ylabel(config.YLabel.value());
+    }
+
+    show();
+}
```
Comment (Mr-S-Mirzoev) [inline source/modules/plotter/inc/plotter_impl.h]: trailing newline

File: source/modules/rnd/src/bm.cpp [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    // Linear interpolation
+    auto t_i = linalg::range(0, T, N);
+    return linalg::interp(t_i, BM, t);
+}
+
+/**
+ * Function that returns the linear interpolation WInt of the values
+ * \f$W_{t_ 0},W_{t_ 1},...,W_{T}\f$ with \f$t_i = i*T/N\f$ at the point \f$t \in [0, T]\f$
+*/
```
Comment (Mr-S-Mirzoev) [inline source/modules/rnd/src/bm.cpp]: move doc-string to header

**Review by MaksimSerebryakov — APPROVED** [review]

File: source/modules/linalg/inc/linalg/matrix.h (hunk 10 above)
Comment (Mr-S-Mirzoev) [reply source/modules/linalg/inc/linalg/matrix.h:27]: > what do you think about renaming this method?
> 
> it was a bit difficult for me to understand the meaning before checking matrix.cpp
> 
> maybe smth with "transpose" or if T is common for matrix, it's okey :)

It definitely needs to be commented better, but I guess naming is more or less okay. Yep, it's common. Afaik numpy uses the same operator

Comment (Mr-S-Mirzoev) [reply source/modules/linalg/inc/linalg/matrix.h:27]: BTW it's not that common. But I guess I'll leave it as it is for now but I'll keep in mind that naming might need changing

File: CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,66 @@
+cmake_minimum_required (VERSION 3.0)
+project (num_stoch_calc)
+
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "Default C++ standard")
+set(NUM_STOCH_CALC_EXAMPLE_EXE num_stoch_calc_exe)
+set(NUM_STOCH_CALC_ROOT ${CMAKE_CURRENT_LIST_DIR})
+
+function(SETUP_MODULE MODULE_NAME_LOWERCASE)
```
Comment (Mr-S-Mirzoev) [inline CMakeLists.txt]: Move functions to utils.cmake file

File: source/main.cpp (hunk 13 above)
Comment (Mr-S-Mirzoev) [reply source/main.cpp]: Looks like mpp doesn't support latex

**Pull request merged**